### ISSUE TITLE
Tensorflow unsupported within MSYS2?

### ISSUE THREAD

**Issue by jdpipe** [tensorflow/tensorflow#54284]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): wanted binary .whl via `python3 -m pip install tensorflow`
- TensorFlow version: any would be fine, was trying for 2.6.3
- Python version: 3.9
- Installed using virtualenv? pip? conda?: pip
- CUDA/cuDNN version: no GPU

**Describe the problem**

On MSYS2, using pip via the mingw64-compiled Python, the message is
```
# python3 -m pip install --upgrade https://files.pythonhosted.org/packages/f2/ff/32e0f1e039fb6cabc88a5e03a0a2b2f01cd50b69cc1b915daf80ca0345c5/tensorflow-2.6.3-cp39-cp39-win_amd64.whl
ERROR: tensorflow-2.6.3-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.
```

It's the same if I just use the name eg `python3 -m pip install tensorflow`.

I have an extensive codebase that currently depends on MSYS2 and gcc in the Windows context, and we have a module that successfully uses Tensorflow on our Linux version of the code. As above, however, I can't use Tensorflow via Python under MSYS2 using the version Python that MSYS2 provides.

I wonder if I could get some clarity on what the options are for Windows/MSYS2 therefore:

- can tensorflow be built on Windows using gcc in MSYS2? The instructions, even if they use MSYS2, seem to be referring to MSVC as the compilier. What are the reasons why 'full' MSYS2 support is not possible? *
- should I instead switch, if possible, to building my software to link against Windows Python instead of MSYS2 Python. This would be less convenient for users, but might work.
- would it be possible for the situation regarding MSYS2 support to be clarified in the MSYS2 documentation somehow?

[*] I note that https://www.tensorflow.org/install/source_windows says that there is some issue with path support in Windows by Bazel, but obviously Tensorflow builds find on Linux, where such paths exist. Is this a Bazel configuration issue?

**Comment by tilakrayal** [comment]
@jdpipe ,
Can you please confirm is there any specific reason to use tf v2.6.3, please try with stable v2.7 and let us know if you are facing same issue.Thanks!

**Comment by jdpipe** [comment]
Hi @tilakrayal,
I tried the following instead:
```
pye@DESKTOP-6ADQVP0 MINGW64 ~/solartherm
# python3 -m pip install --upgrade "https://files.pythonhosted.org/packages/dd/0b/6fd33732d436a6735094b6e44359e13c10901e71aeaf348ff7c9e4ee08c9/tensorflow-2.8.0-cp39-cp39-win_amd64.whl"
ERROR: tensorflow-2.8.0-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.
```
and also 
```
pye@DESKTOP-6ADQVP0 MINGW64 ~/solartherm
# python3 -m pip install --upgrade "https://files.pythonhosted.org/packages/af/e0/d1e53ca7411b0dbe8d82afafd39aab5aafaa02fa5d22c3c226a61f349f19/tensorflow-2.7.1-cp39-cp39-win_amd64.whl"
ERROR: tensorflow-2.7.1-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.
```
and also 
```
pye@DESKTOP-6ADQVP0 MINGW64 ~/solartherm
# python3 -m pip install tensorflow
ERROR: Could not find a version that satisfies the requirement tensorflow (from versions: none)
ERROR: No matching distribution found for tensorflow

pye@DESKTOP-6ADQVP0 MINGW64 ~/solartherm
# which python3
/mingw64/bin/python3
```

Note here that I am using the Python that is provided by MSYS2 (MinGW64), which is not the same as 'regular' CPython for Windows. It seems that the PyPI packages for Tensorflow are being deliberately blocked for use on MSYS2 somehow, and I'm not sure if that's deliberate (nothing in the documentation that I could see) or on purpose (would be nice to know the reason).

**Comment by jdpipe** [comment]
Further to the above, I note that I typed `python3 -m pip debug --verbose` and got the following output:
```
Compatible tags: 33
  cp39-cp39-mingw_x86_64
  cp39-abi3-mingw_x86_64
  cp39-none-mingw_x86_64
  cp38-abi3-mingw_x86_64
  cp37-abi3-mingw_x86_64
  cp36-abi3-mingw_x86_64
  cp35-abi3-mingw_x86_64
  cp34-abi3-mingw_x86_64
  cp33-abi3-mingw_x86_64
  cp32-abi3-mingw_x86_64
  py39-none-mingw_x86_64
  py3-none-mingw_x86_64
  py38-none-mingw_x86_64
  py37-none-mingw_x86_64
  py36-none-mingw_x86_64
  py35-none-mingw_x86_64
  py34-none-mingw_x86_64
  py33-none-mingw_x86_64
  py32-none-mingw_x86_64
  py31-none-mingw_x86_64
  py30-none-mingw_x86_64
  cp39-none-any
  py39-none-any
  py3-none-any
  py38-none-any
  py37-none-any
  py36-none-any
  py35-none-any
  py34-none-any
  py33-none-any
  py32-none-any
  py31-none-any
  py30-none-any
```
Meanwhile, if I type ` python3 -m pip download tensorflow --platform win32 --python-version 3.9 -vv --no-deps` then I get this output:
```
[...]
d72c28c6c6fdda814fd68b957a3d206f4df5/tensorflow-2.8.0-cp37-cp37m-macosx_10_14_x86_64.whl#sha256=fa4a723368d5f748b6f4ec305cf7c26b98e4a6a8c2ce1425f8ae10383a37bcfc (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp37-cp37m-manylinux2010_x86_64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/31/66/d9cd0b850397dbd33f070cc371a183b4903120b1c103419e9bf20568456e/tensorflow-2.8.0-cp37-cp37m-manylinux2010_x86_64.whl#sha256=05fb161c6b2a6c4b8317a703a0a6d7f7aa6b5e3c6ea31bbc4f44ef96b89c3344 (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp37-cp37m-win_amd64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/fe/36/7c7c9f106e3026646aa17d599b817525b139e2870f75b532318573cbabd4/tensorflow-2.8.0-cp37-cp37m-win_amd64.whl#sha256=291fa84f1022914580810ad76732fb254e44a8a609128e1c58873a12b2f81559 (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp38-cp38-macosx_10_14_x86_64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/e9/6b/a4b872afe267184051a716aae58f1f9d8ebb456ecce7edece70a07c15f56/tensorflow-2.8.0-cp38-cp38-macosx_10_14_x86_64.whl#sha256=dd0f9f113ebc21b73fcd349db1629e187b8686395b8146d100eb1706a943bbc0 (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp38-cp38-manylinux2010_x86_64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/2f/45/f5c91d69c2121e8e60673164bdcd2c6cda7b89e37decbc3c01b0466ca990/tensorflow-2.8.0-cp38-cp38-manylinux2010_x86_64.whl#sha256=8489b4f1771e146f752b0eaeb57acf183bd07357e4550464e7dff18b3b656b5d (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp38-cp38-win_amd64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/17/5b/dac6e0607e4186b9e157597cd96d945aa769c60ef9f9f1b7ddc174f39332/tensorflow-2.8.0-cp38-cp38-win_amd64.whl#sha256=da38d4043185267e7316ae5dc98d18e89c8af4170859f64798e7a3607fd606e3 (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp39-cp39-macosx_10_14_x86_64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/80/26/2990ee24ea5ef5ece557828ce295abc75e5b6c0c8ebf571a842f6e1125db/tensorflow-2.8.0-cp39-cp39-macosx_10_14_x86_64.whl#sha256=52f225fecc688281b3ae2cba2b52d3ed6215ed4a3ffb686b9cfd09885ca65563 (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp39-cp39-manylinux2010_x86_64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/b9/39/88799b7257f73482d4c77bc18c068902a91629c6d380db88cc28ede44fd5/tensorflow-2.8.0-cp39-cp39-manylinux2010_x86_64.whl#sha256=9d91a989e5455ae713c03fd7236071ab3f232ad8ff2831f2658072933546091f (from https://pypi.org/simple/tensorflow/)
  Skipping link: none of the wheel's tags (cp39-cp39-win_amd64) are compatible (run pip debug --verbose to show compatible tags): https://files.pythonhosted.org/packages/dd/0b/6fd33732d436a6735094b6e44359e13c10901e71aeaf348ff7c9e4ee08c9/tensorflow-2.8.0-cp39-cp39-win_amd64.whl#sha256=b360c13b3e58b9a5c0780cbdb6b549eea73f620275fa203f8508fe418ae02735 (from https://pypi.org/simple/tensorflow/)
[...]
```
Which clearly indicates that it is the 'platform tag' that is causing the problem here. In short, `cp39-cp39-win_amd64` doesn't match with `cp39-cp39-mingw_x86_64`. 

The question is, does the platform need to be so restrictive here? If Python/Wheel wasn't blocking me, would this module actually work, or not? Is this a MSYS2 issue or a TensorFlow issue, I wonder?

https://www.python.org/dev/peps/pep-0425/#platform-tag

**Comment by jdpipe** [comment]
One more thing. If I manually unzip the .whl file
```
wget "https://files.pythonhosted.org/packages/dd/0b/6fd33732d436a6735094b6e44359e13c10901e71aeaf348ff7c9e4ee08c9/tensorflow-2.8.0-cp39-cp39-win_amd64.whl"
mkdir ~/tmpwhl
cd ~/tmpwhl
bsdtar xvf ../tensorflow-2.8.0-cp39-cp39-win_amd64.whl
```
and then 'trick' Python into importing that package:
```
pye@DESKTOP-6ADQVP0 MINGW64 ~/tmpwhl
# PYTHONPATH=. ipython3
Python 3.9.7 (default, Nov 21 2021, 22:02:56)  [GCC 11.2.0 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
C:/msys64/home/pye/tmpwhl/tensorflow/python/pywrap_tensorflow.py in <module>
     63   try:
---> 64     from tensorflow.python._pywrap_tensorflow_internal import *
     65   # This try catch logic is because there is no bazel equivalent for py_extension.

ImportError: DLL load failed while importing _pywrap_tensorflow_internal: The specified module could not be found.

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-d6579f534729> in <module>
----> 1 import tensorflow

C:/msys64/home/pye/tmpwhl/tensorflow/__init__.py in <module>
     39 import sys as _sys
     40
---> 41 from tensorflow.python.tools import module_util as _module_util
     42 from tensorflow.python.util.lazy_loader import LazyLoader as _LazyLoader
     43

C:/msys64/home/pye/tmpwhl/tensorflow/python/__init__.py in <module>
     38 # pylint: disable=wildcard-import,g-bad-import-order,g-import-not-at-top
     39
---> 40 from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow
     41 from tensorflow.python.eager import context
     42

C:/msys64/home/pye/tmpwhl/tensorflow/python/pywrap_tensorflow.py in <module>
     77     sys.setdlopenflags(_default_dlopen_flags)
     78 except ImportError:
---> 79   raise ImportError(
     80       f'{traceback.format_exc()}'
     81       f'\n\nFailed to load the native TensorFlow runtime.\n'

ImportError: Traceback (most recent call last):
  File "C:/msys64/home/pye/tmpwhl/tensorflow/python/pywrap_tensorflow.py", line 64, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: The specified module could not be found.


Failed to load the native TensorFlow runtime.
See https://www.tensorflow.org/install/errors for some common causes and solutions.
If you need help, create an issue at https://github.com/tensorflow/tensorflow/issues and include the entire stack trace above this error message.

In [2]:
```
I opened `_pywrap_tensorflow_internal.pyd` using Dependency Walker, and found that the file that was missing was `PYTHON39.DLL`. On MinGW64, this file is called `libpython3.9.dll`, so I made a copy of the MinGW64 file, and renamed it to the expected name `python39.dll`.

I then made a tiny bit more progress, namely:
```
pye@DESKTOP-6ADQVP0 MINGW64 ~/tmpwhl
# ipython3
Python 3.9.7 (default, Nov 21 2021, 22:02:56)  [GCC 11.2.0 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow
2022-02-13 07:08:14.285665: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2022-02-13 07:08:14.400592: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
```
This looks like it's getting a little bit further, and attempting to `dlopen` some CUDA-related stuff, and my system doesn't have CUDA, so I guess it fails. But presumably, it should not be crashing out of Python in this case.

All in all, it looks like the `mingw_x86_64` 'platform tag' makes sense, because the DLL for Python has a different name on this platform compared to `win_amd64`, and that means that platform-sensitive Python modules need to be compiled differently on this platform. 

Having established that, are there any suggestions for how to compile Tensorflow such that it works with the MSYS2-MinGW64 version of Python?

**Comment by tilakrayal** [comment]
@jdpipe ,
Can you please take a look at this [link](https://github.com/tensorflow/tensorflow/issues/36167#issuecomment-577886156) for the issue with the similar DLL error.It helps.Thanks!

**Comment by SuryanarayanaY** [comment]
Hi @jdpipe ,

The prebuilt binaries for windows are of win_amd64.whl type and seems not supporting for mingw_x86_64 architecture.

You have option for Build from source using MSYS shell. Please refer to the documentation here for [same](https://www.tensorflow.org/install/source_windows#build_using_the_msys_shell). May be this can be helpful for your case.
Please check and confirm if this solves your purpose. Thank you!

**Comment by jdpipe** [comment]
Hi @SuryanarayanaY,
Thanks for the documentation link. It says that one can build using the MSYS2 bash shell.
But would this produce a version of Tensorflow that is built using MINGW's GCC and links to the MINGW version of Python? 
The instructions say to use the MSYS2 shell, but then all of the instructions for installing dependencies, specifically MS Visual C++ Build Tools, etc. (ie "then follow the previous instructions for the Windows native command line").
I feel that the instructions are not really sufficiently clear. For example, if using MSYS2, and I want to use the MSYS2 version of Python, why would I install Windows Python? It would be helpful to have an end-to-end set of instructions for building on MSYS2, using native MSYS2 GCC and Python, etc. 
It would be even better if PyPI could host the resulting packages ;-) But as noted, there may be issues with CUDA support in MSYS2 (not a big deal for me, but certainly it would be for others).

**Comment by cminnoy** [comment]
I'm in a similar situation. I need to adapt an open source application that was build under Windows using MYSYS2. Now we want to add a tensorflow model to the application. Two things are now bothering us. First tensorflow doesn't work under MYSYS2 (at least we can't get it to build/work). Secondly, From Tensorflow 2.11 GPU support is only under WSL2, but not anymore under Windows directly. GPU accelerated computing is really crucial.

**Comment by mihaimaruseac** [comment]
Most likely this should be closed as duplicate of #36167
